README.markdown
Last Updated : 2012/05/29 17:52:13 tkych


# Donuts: Graph DSL for Common Lisp


## Introduction

By donuts, the graph represented by s-expression is converted to image.
How to use donuts is easy.
<> creates a node.
-> puts an edge between two nodes.
&& makes a graph by bundling some nodes, edges and graphs.
$$ outputs an image of the graph.

For more details, please see index(Under Translation) or index-ja(Japanease) in doc directory.


## The Goal of Donuts

[Graphviz][] is a collection of libraries and utilities for drawing a graph.
[Dot language][] is description language, used in Graphviz.
Graphviz is very useful.
However, I (as a lisp programmer) think there are some points to do kaizen.

1.  Since dot language is not turing-complete,
    when we draw a graph, we don't take full advantage of the pattern in the graph.

2.  Because dot language is so-called compiled language,
    development cycle is inconvenient.

3.  Plain common lisp does not have ability to draw graph.

The goal of donuts is to draw graph in lispic way of thinking 
(REPL, macro, CLOS, multi-paradigm style, and so on). 

  [Graphviz]: http://www.graphviz.org/
  [Dot language]: http://www.graphviz.org/dot-language.html


## Current Version of Donuts

0.3.1 (beta)


## Dependencies

* [Graphviz][] by AT&T Labs

* [CL-PPCRE](http://weitz.de/cl-ppcre/) by Dr. Edmund Weitz

* [Trivial-Shell](http://common-lisp.net/project/trivial-shell/) by Gary Warren King


## Installation & Start

1.  CL-REPL> `(ql:quickload :donuts)`
2.  CL-USER> `(in-package :donuts)`
3.  DONUTS> `(dot-output (&& (-> 1 2)))`  ;output dot code in standard-output
4.  DONUTS> `($$ (&& (-> 1 2)))`  ;output graph image to viewer


## Usage

* Node-Constructor, <> makes node from node's identity.   `(<> label) => node`
* Edge-Constructor, -> makes edge between nodes.   `(-> node1 node2) => edge`
* Graph-Constructor, &&, &, [&] makes graph with nodes, edges, graphs.   `(&& . nodes-edges-graphs) => graph`
* Shell-Interface, $$, $ outputs graph to viewer.   `($$ graph) => NIL ;output image to viewer`
* DOT-OUTPUT outputs dot code in standart-output.   `(DOT-OUTPUT graph) => NIL ;output dot code`


## Examples

      DONUTS> (dot-output
                (& (:label "example")
                  (-> (<> "a" :shape :box) "b" :color :red)))

      digraph graph_ID_41 {
        label="example";
        node_ID_39 [label="a",shape=box];
        node_ID_39 -> "b" [color=red];
      }
      NIL

      DONUTS> ;; Example from http://graphviz.org/content/cluster
              ($ (:outfile "cluster.pdf")
                 (&& ([&] (:label "process #1" :style :filled :color :lightgrey)
                       (with-node (:style :filled :color :white)
                         (--> "a0" "a1" "a2" "a3")))
                     ([&] (:label "process #2" :color :blue)
                       (with-node (:style :filled)
                         (--> "b0" "b1" "b2" "b3")))
                     (->> (<> "start" :shape :Mdiamond) "a0" "b0")
                     (==> "a3" "b3" (<> "end" :shape :Msquare))
                     (->  "a1" "b3")
                     (->  "a3" "a0")
                     (->  "b2" "a3")))

       ; Create cluster.pdf & Output image to viewer as shown on the right
       NIL

       DONUTS> 
       ;; Example from http://www.linuxjournal.com/article/7275
       ;; num-day: total number of days in month
       ;; starting-day: 0 as Sun, 1 as Mon, ... , 6 as Sat
       (defun generate-monthly-calendar (month year num-days starting-day)
         (let ((month (<> (format nil "~@(~A~)\\n~D" month year) :shape :Msquare))
               (luminary7 (loop :for day :in '("Sun" "Mon" "Tue" "Wed" "Thu" "Fri" "Sat")
                                :collect (<> day :shape :egg :style :filled :color :lightgray)))
               (days (loop :for day :in (nconc (loop :repeat starting-day :collect "")
                                               (loop :for d :from 1 :to num-days :collect d)
                                               (loop :repeat (- (* 7 (if (and (= 28 num-days) (= 0 starting-day))
                                                                         4 5)) ;for Feb starting Sun in common year 
                                                                starting-day num-days)
                                                     :collect ""))
                           :collect (<> day :shape :box))))
           ;; fn group from On Lisp, ex. (group '(1 2 3 4) 2) => ((1 2) (3 4))
           (apply #'&& (loop :for week :in (cons luminary7 (group days 7)) 
                             :collect (apply #'--> month week)))))

       GENERATE-MONTHLY-CALENDAR

       DONUTS> ($$ (& (:size "8,6":rankdir :LR)
                     (generate-monthly-calendar 'may 2012 31 2)))

       ; Output Calendar to Viewer
       NIL

       DONUTS> 
       ;; Example from http://www.graphviz.org/doc/info/html2.gv
       ($$ (& (:rankdir :LR)
         (with-node (:shape :plaintext)
           (let ((a (<> (html (table :border 0 :cellborder 1 :cellspacing 0
                                     (tr (td :rowspan 3 :bgcolor :yellow "class"))
                                     (tr (td :port "here" :bgcolor :lightblue "qualfier"))))))
                 (b (<> (html (table :bgcolor :bisque
                                     (tr (td :colspan 3 "elephant")
                                         (td :rowspan 2 :bgcolor :chartreuse
                                             :valign :bottom :align :right "two"))
                                     (tr (td :colspan 2 :rowspan 2
                                             (table :bgcolor :grey
                                                    (tr (td "corn"))
                                                    (tr (td :bgcolor :yellow "c"))
                                                    (tr (td "f"))))
                                         (td :bgcolor :white "penguin"))
                                     (tr (td :colspan 2 :border 4 :align :right :port "there" "4"))))
                         :shape :ellipse :style :filled))
                 (c (<> (html "long line 1" (br) "line 2" (br :align :left) "line 3" (br :align :right))))
                 (d (<> "d" :shape :triangle)))
             (&&
               (~ b c)
               (-> (@ a :here) (@ b :there) :dir :both :arrowtail :diamond)
               (-> c b)
               (-> d c :label (html (table (tr (td :bgcolor :red :width 10)
                                               (td "Edge labels" (br) "also")
                                               (td :bgcolor :blue :width 10))))))))))

       ; Output example to Viewer
       NIL


## Author, License, Copyright

* Takaya OCHIAI <tkych.repl@gmail.com>

* MIT License

* Copyright (C) 2012 Takaya OCHIAI


CHANGELOG
Version 0.3.1
2012-05-21
Fix Bug: macro WITH-NODE, WITH-EDGE.

Version 0.3.0
2012-05-18
Modify /SRC: rm nuts.lisp, fresh.lisp.
           : add NODE.LISP, EDGE.LISP, GRAPH.LISP, DOT-OUTPUT.LISP, SESAME.LISP, DONUTS-UTILS.LISP.
Del   Macro: ->>, ==>, -<, >-, <==.
Del   Fn   : <-.
Add   Fn   : ->>, ==>, -<.

Version 0.2.8
2012-05-13
Fix Bug: macro HTML in src/html-like-labels.lisp.

Version 0.2.7
2012-05-10
Add: src/HTML-LIKE-LABELS.LISP.

Version 0.2.6
2012-05-08
Initial release.

src/dot-output.lisp
;;;; Last Updated : 2012/05/21 17:45:08 tkych

;; dot-output.lisp in donuts/src/

;; Copyright (c) 2012 Takaya OCHIAI

;; Permission is hereby granted, free of charge, to any person
;; obtaining a copy of this software and associated documentation
;; files (the "Software"), to deal in the Software without
;; restriction, including without limitation the rights to use, copy,
;; modify, merge, publish, distribute, sublicense, and/or sell copies
;; of the Software, and to permit persons to whom the Software is
;; furnished to do so, subject to the following conditions:

;; The above copyright notice and this permission notice shall be
;; included in all copies or substantial portions of the Software.

;; THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
;; EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
;; MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
;; NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
;; BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
;; ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
;; CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
;; SOFTWARE.

;;====================================================================
;; Dot-Output
;;====================================================================
(in-package :in-donuts)

(defparameter *declared* nil)
(defun declared? (node) (member node *declared*))

(defun dot-output (graph)
  (unless (graph? graph) (error "~A isn't graph type." graph))
  (with-slots (name attrs buff dir cache strict) graph
     (~&) (when strict (princ "strict "))
     (format t "~A ~A {~%" (if dir "digraph" "graph") name)
     (awhen attrs
       (format t "~&~{  ~A=~A;~^~&~}~&" (escape-attrs it)))
     (aif cache
          (princ it)
          (let ((output
                 (with-output-to-string (s)
                   (let ((*standard-output* s) (*declared* nil))
                     (output-buff buff)
                     (setf buff nil))))) ;for gc
            (setf cache output)
            (princ output)))
     (format t "~&}")))

(defun output-buff (buff)
  (dolist (x buff)
    (cond ((edge? x)     (output-edge x))
          ((graph? x)    (output-subgraph x))
          ((pre-node? x) (format t "~&  ~S;" x))
          ((node? x)     (declare-node x))
          ((sesame? x)   (output-sesame x)) ;for with
          ((null x)      nil)               ;for null graph
          (t             (error "~A is not graph content type." x)))))

;;--------------------------------------
(defun declare-node (node)
  (when (and (node? node) (not (declared? node)))
    (format t "~&  ~A [~{~A=~A~^,~}];"
            (:name node) (escape-attrs (:attrs node)))
    (push node *declared*)))

(defun output-node (node)
  (cond ((pre-node? node) (format t "~S" node))
        ((sesame? node)   (format t "~A" (open-sesame node))) ;for port
        ((node? node)     (format t "~A" (:name node)))
        (t                (error "~A is not node type." node))))

;;--------------------------------------
(defgeneric output-edge (edge))

(defmethod output-edge :around ((edge edge))
  (with-slots (context) edge
     (if (or (null context)
             (equal context *with-edge-context*)) ;!!!!
         (call-next-method edge)
         (progn
           (format t "~&  { edge [~{~A=~A~^,~}];"
                   (escape-attrs context))
           (call-next-method edge)
           (format t "~&  };")))))

(defmethod output-edge ((edge normal-edge))
  (with-slots (node1 node2 attrs) edge
     (declare-node node1)
     (declare-node node2)
     (~&) (princ "  ")
     (output-node node1)
     (if (path? edge) (princ " -- ") (princ " -> "))
     (output-node node2)
     (when attrs
       (format t " [~{~A=~A~^,~}]" (escape-attrs attrs)))
     (princ ";")))

(defmethod output-edge ((edge penetrate-edge))
  (with-slots (nodes) edge
     (dolist (node nodes) (declare-node node))
     (~&) (princ "  ")
     (output-node (1st nodes))
     (dolist (node (rest nodes))
       (if (path? edge) (princ " -- ") (princ " -> "))
       (output-node node))
     (princ ";")))

(defmethod output-edge ((edge radiate-edge))
  (with-slots (origin-node nodes) edge
     (declare-node origin-node)
     (dolist (node nodes) (declare-node node))
     (~&) (princ "  ")
     (output-node origin-node)
     (if (path? edge) (princ " -- {") (princ " -> {"))
     (output-node (1st nodes))
     (dolist (node (rest nodes))
       (princ "; ")
       (output-node node))
     (princ "};")))

(defmethod output-edge ((edge converge-edge))
  (with-slots (nodes converge-node) edge
     (declare-node converge-node)
     (dolist (node nodes) (declare-node node))
     (dolist (node nodes)
      (~&) (princ "  ")
      (output-node node)
      (if (path? edge) (princ " -- ") (princ " -> "))
      (output-node converge-node)
      (princ ";"))))

;;--------------------------------------
(defgeneric output-subgraph (graph))

(defmethod output-subgraph ((graph graph))
  (with-slots (buff cache) graph
     (if cache
         (format t "~A" cache)
         (output-buff buff))))

(defmethod output-subgraph ((cluster cluster))
  (with-slots (name attrs buff cache) cluster
     (format t "~&  subgraph ~A {~%" name)
     (awhen attrs
       (format t "~&~{  ~A=~A;~^~&~}~&" (escape-attrs it)))
     (if cache
         (format t "~A~&  }" cache)
         (progn (output-buff buff)
                (format t "~&  }")))))

;;--------------------------------------
(defparameter *capital-vals* '(:Mdiamond :Msquare :Mcircle :Mrecord))
(defun capital-val? (x) (member x *capital-vals*))

(defparameter *upper-vals*
              '(:LR :RL :TB :BT :BL :BR :TL :TR :RB :RT :LB :LT))
(defun upper-val? (x) (member x *upper-vals*))

(defparameter *URL-attrs* '(:URL :headURL :tailURL :labelURL :edgeURL))
(defun URL-attr? (x) (member x *URL-attrs*))
(defun escape-url-attr (url-attr)
  (format nil "~{~(~A~)~}URL"
          (aif (ppcre:split "URL" (symbol-name url-attr))
               it '(""))))

(defun escape-attrs (alst)
  (let ((len (length alst)) (acc nil))
    (do ((i 0 (1+ i)))
        ((<= len i) (nreverse acc))
      (let ((x (nth i alst)))
        (push
         (if (evenp i)
             (if (URL-attr? x)
                 (escape-url-attr x)
                 (string-downcase (symbol-name x)))
             (cond ((sesame? x)  (open-sesame x)) ;for html-like-label
                   ((stringp x)  (str "\"" x "\""))
                   ((numberp x)  x)
                   ((eql x t)    "true")
                   ((eql x nil)  "false")
                   ((keywordp x)
                    (cond ((upper-val? x)   (symbol-name x))
                          ((capital-val? x) (string-capitalize
                                             (symbol-name x)))
                          (t                (string-downcase
                                             (symbol-name x)))))
                   ((cluster? x) (:name x))
                   (t (error "~A is not an attribute value type." x))))
         acc)))))

;;--------------------------------------
(defun output-sesame (sesame)
  (format t "~&~A" (open-sesame sesame)))

;;====================================================================